core/commands/swarm.go

```diff
@@ -557,10 +560,43 @@ func parseAddresses(addrs []string) (iaddrs []iaddr.IPFSAddr, err error) {
 	return
 }
 
+// parseMultiaddrs is a function that takes in a slice of peer multiaddr
+// and returns slices of multiaddrs and peerids
+func parseMultiaddrs(maddrs []ma.Multiaddr) (iaddrs []iaddr.IPFSAddr, err error) {
+	iaddrs = make([]iaddr.IPFSAddr, len(maddrs))
+	for i, maddr := range maddrs {
+		iaddrs[i], err = iaddr.ParseMultiaddr(maddr)
+		if err != nil {
+			return nil, cmds.ClientError("invalid peer address: " + err.Error())
+		}
+	}
+	return
+}
+
 // peersWithAddresses is a function that takes in a slice of string peer addresses
 // (multiaddr + peerid) and returns a slice of properly constructed peers
 func peersWithAddresses(addrs []string) ([]pstore.PeerInfo, error) {
-	iaddrs, err := parseAddresses(addrs)
+	// resolve the addrs
+
+	var maddrs []ma.Multiaddr
+	for _, addr := range addrs {
+		maddr, err := ma.NewMultiaddr(addr)
+		if err != nil {
+			return nil, err
+		}
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		raddrs, err := madns.Resolve(ctx, maddr)
+		cancel()
+		if err != nil {
+			return nil, err
+		}
+		if len(raddrs) == 0 {
+			return nil, fmt.Errorf("non-resolvable multiaddr about %v", maddr)
+		}
+		maddrs = append(maddrs, raddrs...)
```

Comment:
I believe we may need to filter out addresses that *still* don't end in `/ipfs/Qm...`.